### PR TITLE
Allow more attributes in `clippy::useless_attribute`

### LINKED ISSUE #4467: Incorrect suggestion in `useless_attribute` lint
The `useless_attribute` lint is meant to warn about `extern crate` and `use` items with outer lint attributes (for example `#[allow(dead_code)]`).

However, when this lint triggers on items that are preceded by other items, applying the suggestion will fail to compile.

Minimal example ([Playground Link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=048a33793c45d6a4238a6d73ab940db5)):

```rust
#![allow(unused_imports)]

fn main() {}

#[allow(dead_code)]
use aho_corasick;
```

Produces:

```
error: useless lint attribute
 --> src/main.rs:6:1
  |
6 | #[allow(dead_code)]
  | ^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![allow(dead_code)]`
```

But applying the suggestion results in:

```
error: an inner attribute is not permitted in this context
 --> src/main.rs:6:3
  |
6 | #![allow(dead_code)]
  |   ^
```

To fix this, we could

a) Try to find a way to detect if the `use`/`extern crate` is preceded by any other items
b) Remove the suggestion completely if a) turns out impossible

### LINKED ISSUE #10878: Incorrect `clippy::useless_attribute` for `ambiguous_glob_reexports`
### Summary

The new stable (rust 1.70) `ambiguous_glob_reexports` is valid in use items.

### Lint Name

useless_attribute

### Reproducer

I tried this code:

```rust
mod foo1 {
    pub struct Foo;
}

mod foo2 {
    pub struct Foo;
}

// #[allow(clippy::useless_attribute)]
#[allow(ambiguous_glob_reexports)]
pub use foo1::*;
pub use foo2::*;
```

I saw this happen:

Deny error in `#[allow(ambiguous_glob_reexports)]`.

I expected to see this happen:

No error

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-pc-windows-msvc
release: 1.70.0
LLVM version: 16.0.2
```

### Additional Labels

_No response_

### LINKED ISSUE #11595: Incorrect `clippy::useless_attribute` for `hidden_glob_reexports`
### Summary

`hidden_glob_reexports` is valid on `use` items.

### Lint Name

useless_attribute

### Reproducer

I tried this code:

```rust
#![allow(dead_code)]
#![allow(unused_imports)]

mod foo1 {
    pub struct Foo;
    pub use Foo as Bar;
}

mod foo2 {
    pub struct Foo;
}

pub use foo1::*;

// #[allow(clippy::useless_attribute)]
#[allow(hidden_glob_reexports)]
use foo2::Foo as Bar;
```

I saw this happen:

```
error: useless lint attribute
  --> src/lib.rs:16:1
   |
16 | #[allow(hidden_glob_reexports)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![allow(hidden_glob_reexports)]`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_attribute
   = note: `#[deny(clippy::useless_attribute)]` on by default

error: could not compile `ice` (lib) due to previous error
```

I expected to see this happen: No errors because `hidden_glob_reexports` lint would be generated without it:

```
warning: private item shadows public glob re-export
  --> src/lib.rs:17:5
   |
17 | use foo2::Foo as Bar;
   |     ^^^^^^^^^^^^^^^^
   |
note: the name `Bar` in the type namespace is supposed to be publicly re-exported here
  --> src/lib.rs:13:9
   |
13 | pub use foo1::*;
   |         ^^^^^^^
note: but the private item here shadows it
  --> src/lib.rs:17:5
   |
17 | use foo2::Foo as Bar;
   |     ^^^^^^^^^^^^^^^^
   = note: `#[warn(hidden_glob_reexports)]` on by default
```

### Version

```text
rustc 1.75.0-nightly (e0d7ed1f4 2023-10-01)
binary: rustc
commit-hash: e0d7ed1f453fb54578cc96dfea859b0e7be15016
commit-date: 2023-10-01
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.2
```

### Additional Labels

_No response_

### LINKED ISSUE #12753: Incorrect `clippy::useless_attribute` for `disallowed_types`
### Summary

`disallowed_types` is valid on `use` items.

### Lint Name

useless_attribute

### Reproducer

With this `clippy.toml`:
```toml
[[disallowed-types]]
path = "std::io::Stdout"
reason = "Use my_crate::StdoutSafe"
```

I tried this code:

```rust
#[allow(clippy::disallowed_types)]
use std::io::Stdout;
```

I saw this happen:

```
error: useless lint attribute
 --> src/stdout.rs:5:1
  |
5 | #[allow(clippy::disallowed_types)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![allow(clippy::disallowed_types)]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_attribute
  = note: `#[deny(clippy::useless_attribute)]` on by default
```

I expected to see this happen: (no failing lints)

### Version

```text
rustc 1.77.2 (25ef9e3d8 2024-04-09)
binary: rustc
commit-hash: 25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04
commit-date: 2024-04-09
host: aarch64-apple-darwin
release: 1.77.2
LLVM version: 17.0.6
```

### Additional Labels

@rustbot label +C-bug
@rustbot label +I-false-positive

### PR DESCRIPTION
**Pull request by 9999years**
Fixes #12753 
Fixes #4467 
Fixes #11595 
Fixes #10878

changelog: [`useless_attribute`]: Attributes allowed on `use` items now include `ambiguous_glob_exports`, `hidden_glob_reexports`, `dead_code`, `unused_braces`, and `clippy::disallowed_types`.

### CODE REVIEW

File: clippy_lints/src/attrs/useless_attribute.rs [hunk 1]
```diff
@@ -25,6 +25,10 @@ pub(super) fn check(cx: &LateContext<'_>, item: &Item<'_>, attrs: &[Attribute])
                                 || is_word(lint, sym!(unreachable_pub))
                                 || is_word(lint, sym!(unused))
                                 || is_word(lint, sym!(unused_import_braces))
+                                || is_word(lint, sym!(unused_braces))
+                                || is_word(lint, sym!(dead_code))
```
Comment (xFrednet) [inline clippy_lints/src/attrs/useless_attribute.rs:29]: The CI is currently failing due to this:
```suggestion
                                || is_word(lint, rustc_span::sym::dead_code)
```

**Comment by xFrednet** [comment]
Thank you for the PR, I'll take a look at it soon.

@GuillaumeGomez If you have the time, you could do a prereview :D

**Comment by GuillaumeGomez** [comment]
Sure!

File: clippy_lints/src/attrs/useless_attribute.rs [hunk 2]
```diff
@@ -25,6 +25,10 @@ pub(super) fn check(cx: &LateContext<'_>, item: &Item<'_>, attrs: &[Attribute])
                                 || is_word(lint, sym!(unreachable_pub))
                                 || is_word(lint, sym!(unused))
                                 || is_word(lint, sym!(unused_import_braces))
+                                || is_word(lint, sym!(unused_braces))
```
Comment (GuillaumeGomez) [inline clippy_lints/src/attrs/useless_attribute.rs:28]: At this point, shouldn't we do something like:

```rust
if lint.si_word() && let Some(ident) = lint.ident() && matches!(ident.name, sym::deprecated | sym!(unreachable_pub) | ...)
```

?

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Other than my suggestion, looks all good to me, thanks!

File: clippy_lints/src/attrs/useless_attribute.rs (hunk 2 above)
Comment (9999years) [reply clippy_lints/src/attrs/useless_attribute.rs:28]: RIP:

```rust
error[E0164]: expected tuple struct or tuple variant, found associated function `rustc_span::symbol::Symbol::intern`
  --> clippy_lints/src/attrs/useless_attribute.rs:29:37
   |
29 | ...                   sym!(ambiguous_glob_reexports)
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `fn` calls are not allowed in patterns
   |
   = help: for more information, visit https://doc.rust-lang.org/book/ch18-00-patterns.html
   = note: this error originates in the macro `sym` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Comment (GuillaumeGomez) [reply clippy_lints/src/attrs/useless_attribute.rs:28]: Then we can use `ident.name.as_str()` and match on strings. :)

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Looks good to me, thank you for the changes!

@bors r=xFrednet,GuillaumeGomez

**Pull request merged**